libvast/vast/hash_index.hpp

```diff
@@ -0,0 +1,256 @@
+      ++i;
+    } while (i < max_hash_rounds);
+    return caf::none;
+  }
+
+  /// Locates the digest for a given input.
+  key find_digest(data_view x) const {
+    auto i = find_seed(x);
+    return key{i != seeds_.end() ? hash(x, i->second) : hash(x, 0)};
+  }
+
+  bool append_impl(data_view x, id) override {
+    // After we deserialize the index, we can no longer append data.
+    if (immutable())
+      return false;
+    auto digest = make_digest(x);
+    if (!digest)
+      return false;
+    digests_.push_back(digest->bytes);
+    return true;
+  }
+
+  caf::expected<ids>
+  lookup_impl(relational_operator op, data_view x) const override {
+    VAST_ASSERT(rank(this->mask()) == digests_.size());
+    // Implementation of the one-pass search algorithm that computes the
+    // resulting ID set. The predicate depends on the operator and RHS.
+    auto scan = [&](auto predicate) -> ids {
+      ewah_bitmap result;
+      auto rng = select(this->mask());
+      if (rng.done())
+        return result;
+      for (size_t i = 0, last_match = 0; i < digests_.size(); ++i) {
+        if (predicate(digests_[i])) {
+          auto digests_since_last_match = i - last_match;
+          if (digests_since_last_match > 0)
+            rng.next(digests_since_last_match);
+          result.append_bits(false, rng.get() - result.size());
+          result.append_bit(true);
+          last_match = i;
+        }
+      }
+      return result;
+    };
+    if (op == equal || op == not_equal) {
+      auto k = find_digest(x);
+      auto eq = [=](const digest_type& digest) { return k == digest; };
+      auto ne = [=](const digest_type& digest) { return k != digest; };
+      return op == equal ? scan(eq) : scan(ne);
+    }
+    if (op == in || op == not_in) {
+      // Ensure that the RHS is a list of strings.
+      auto keys = caf::visit(
+        detail::overload([&](auto xs) -> caf::expected<std::vector<key>> {
+          using view_type = decltype(xs);
+          if constexpr (detail::is_any_v<view_type, view<set>, view<vector>>) {
+            std::vector<key> result;
+            result.reserve(xs.size());
+            for (auto x : xs)
+              result.emplace_back(find_digest(x));
+            return result;
+          } else {
+            return make_error(ec::type_clash, "expected set or vector on RHS",
+                              materialize(x));
+          }
+        }),
+        x);
+      if (!keys)
+        return keys.error();
+      // We're good to go with: create the set predicates an run the scan.
+      auto in_pred = [&](const digest_type& digest) {
+        auto cmp = [=](auto& k) { return k == digest; };
+        return std::any_of(keys->begin(), keys->end(), cmp);
+      };
+      auto not_in_pred = [&](const digest_type& digest) {
+        auto cmp = [=](auto& k) { return k == digest; };
+        return std::none_of(keys->begin(), keys->end(), cmp);
+      };
+      return op == in ? scan(in_pred) : scan(not_in_pred);
```

Comment:
I'm not convinced that the optimizer can eliminate the unused lambda instantiation here, since it is capturing variables and `scan` takes the predicate by-value since it needs to be called multiple times.

Instead, I'd go with this:

```cpp
auto is_digest = [=](const auto& key) {
  return key == digest;
};
if (op == in)
  scan([&](const digest_type& digest) {
    return std::any_of(keys->begin(), keys->end(), is_digest);
  });
else
  scan([&](const digest_type& digest) {
    return std::none_of(keys->begin(), keys->end(), is_digest);
  }); 
```

There is a similar case just above.

Reply:
It's one of those instances where I went for easier readability at the negligible (potential!) performance pessimization.

Your solution doesn't work because `digest` is the argument of the predicate that gets captured.

Reply:
This would work, but I find it much less readable:
```diff
diff --git a/libvast/vast/hash_index.hpp b/libvast/vast/hash_index.hpp
index 8fb603d9..1e7a46cc 100644
--- a/libvast/vast/hash_index.hpp
+++ b/libvast/vast/hash_index.hpp
@@ -225,15 +225,19 @@ private:
       if (!keys)
         return keys.error();
       // We're good to go with: create the set predicates an run the scan.
-      auto in_pred = [&](const digest_type& digest) {
-        auto cmp = [=](auto& k) { return k == digest; };
-        return std::any_of(keys->begin(), keys->end(), cmp);
-      };
+      if (op == in) {
+        auto in_pred = [&](const digest_type& digest) {
+          auto cmp = [=](auto& k) { return k == digest; };
+          return std::any_of(keys->begin(), keys->end(), cmp);
+        };
+        return scan(in_pred);
+      }
+      VAST_ASSERT(op == not_in);
       auto not_in_pred = [&](const digest_type& digest) {
         auto cmp = [=](auto& k) { return k == digest; };
         return std::none_of(keys->begin(), keys->end(), cmp);
       };
-      return op == in ? scan(in_pred) : scan(not_in_pred);
+      return scan(not_in_pred);
     }
     return make_error(ec::unsupported_operator, op);
   }
```

Reply:
This also works and is even more readable:

```diff
diff --git a/libvast/vast/hash_index.hpp b/libvast/vast/hash_index.hpp
index fd5305e3..41c5a95c 100644
--- a/libvast/vast/hash_index.hpp
+++ b/libvast/vast/hash_index.hpp
@@ -233,15 +233,12 @@ private:
       if (!keys)
         return keys.error();
       // We're good to go with: create the set predicates an run the scan.
-      auto in_pred = [&](const digest_type& digest) {
+      return scan([&](const digest_type& digest) {
         auto cmp = [=](auto& k) { return k == digest; };
-        return std::any_of(keys->begin(), keys->end(), cmp);
-      };
-      auto not_in_pred = [&](const digest_type& digest) {
-        auto cmp = [=](auto& k) { return k == digest; };
-        return std::none_of(keys->begin(), keys->end(), cmp);
-      };
-      return op == in ? scan(in_pred) : scan(not_in_pred);
+        return op == in 
+          ?  std::any_of(keys->begin(), keys->end(), cmp)
+          : std::none_of(keys->begin(), keys->end(), cmp);
+      });
     }
     return make_error(ec::unsupported_operator, op);
   }
```

Reply:
While it's certainly correct and concise, I'm not sure it it's a performance issues: the predicate we are passing to `scan` gets invoked for every element in the index. The branch predictor will certainly make sure cache misses are minimal, but the extra branching is easy to avoid.